### DISCUSSION TITLE
Integrating Custom Fields into Kiosk Mode

### DISCUSSION THREAD

**Discussion by Does-This-Work** [kimai/Paid-plugins#143] [Questions & Answers]
Hi,

I am exploring ways to integrate the kiosk mode with the custom fields plugin. The goal is to enable employees to add additional information upon logging out. For instance, we want employees to indicate if they have opted for a warm meal provided by the company. My current approach involves using the demo page to test a setup where a custom field, visible only under the activity "Work with meal," is created.

Here’s how I envision it working: an employee goes to a terminal, presses a button to mark the selection of a meal (denoted as "M" in the kiosk field), and then scans their RFID or NFC chip to register their ID alongside this choice. This process aims to track meal consumption for tax purposes. If no meal is taken, the employee would simply scan their chip without pressing the meal button.

However, I’ve encountered a couple of issues:

**1st Problem:** When an employee checks in with the activity labeled as "Work" and checks out with "Work with meal," the system retains the "Work" activity status. I’m currently without a workaround for this. Perhaps introducing separate check-in and check-out activities, or allowing the check-out activity to overwrite the check-in one, could solve this.

**2nd Problem:** The custom field I designed to indicate meal selection does not appear when I start an activity under "Work with meal." However, it becomes visible if I edit the record later. I have set the default value of this field to "yes" as a temporary fix. This issue seems to stem from the form not being dynamic, failing to update based on the selected activity of "Work with meal," despite the field’s visibility settings being restricted to this particular activity. This issue is less problematic in kiosk mode, as users have the option to press a button to indicate whether they had a meal or not.

Image 1: Custom field not appearing when "Work with meal" is initiated.
![image](https://github.com/kevinpapst/kimai-plugins/assets/102190830/094fffae-d1c3-4f33-96e8-945006f57d92)

Image 2: Custom field visible when editing the record.
![image](https://github.com/kevinpapst/kimai-plugins/assets/102190830/061ca9a5-aba4-44f6-974e-09c4081ab638)

To enhance the kiosk further, an addition to the current check-in and check-out screen could be the ability to add custom fields to the table, potentially including other information beyond the date, from, to, and duration.

Image 3: Custom column in kiosk screen (check-in and check-out)
![image](https://github.com/kevinpapst/kimai-plugins/assets/102190830/0d85676e-034e-40f6-b69b-1ff829a66edb)

I’m open to suggestions for improving this setup. We are keen on purchasing these plugins and integrating them in our current test environment, but we want to ensure a smooth integration into our operations before committing. Any input or alternative solutions would be greatly appreciated.

**Comment by pbksol** [comment]
The edit dialog and the available fields are not updated on change of precondition to show a custom field. The field will only show if the entry is saved and then opened again for editing.

There's a note somewhere that it's only working the way it does and that it's impossible to change. But I couldn't find it.

**Reply by Does-This-Work** [reply]
Ahh okay! That isn't much of a problem, because we are planing to use the kiosk mode for 90% of our employees.
Good to know though!

**Comment by kevinpapst** [comment]
I have many improvements for the Kiosk on the roadmap.
But right now the Kiosk mode does not provide a solution for your scenario.

One of my ideas is support for custom-fields and a "checkout form".
But I can't promise anything time-related as of now. 
If you would like to discuss tat contact me via support@kimai.org

**Reply by Does-This-Work** [reply]
Hi,

I tried reaching out to you under that e-mail but I haven't heard back from you.

Best regards

**Reply by kevinpapst** [reply]
I get dozens of emails daily ... send me a reminder now.

**Reply by Does-This-Work** [reply]
Thank you! Sent another email.
